### PR TITLE
CTX-4878: CLI - Basics (user/node config, project select functionality implementation)

### PR DESCRIPTION
**Pull request by bogdant36**
CTX-4878: CLI - Basics (user/node config, project select functionality implementation)

### CODE REVIEW

File: coretex/cli/config.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
+from pathlib import Path
+
+import os
+
+from tabulate import tabulate
+
+import click
+
+from ..configuration import loadConfig, saveConfig, isUserConfigured, isNodeConfigured
+from ..networking.network_manager import NetworkManager
```
Comment (dule1322) [inline coretex/cli/config.py:10]: Why is this used? `networking.networkManager` object should be used to send requests.

File: coretex/cli/config.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            default = True,
+            show_default = False
+        ):
+            return
+
+    click.echo("[Setting up coretex environment]")
+    click.echo("[Node Configuration]")
+
+    nodeName = click.prompt("Machine name", type = str)
+    organizationId = click.prompt("Organization ID", type = str)
```
Comment (dule1322) [inline coretex/cli/config.py:155]: Organization ID is deprecated. Discuss the new flow for node registration with backend team.

File: coretex/cli/config.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+    click.echo("Storage path should be the same as (if) used during --user config")
+    storagePath = click.prompt("Storage path (press enter to use default)", Path.home() / ".coretex", type = str)
+
+    image = click.prompt("Select an image to use for Coretex Node", type = click.Choice(["gpu", "cpu"]), show_choices = True)
+
+    ram = click.prompt("Node RAM memory limit in GB (press enter to use default)", type = int, default = 15)
```
Comment (dule1322) [inline coretex/cli/config.py:162]: `default` should be set to systems available memory in GB.

File: coretex/cli/config.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    swap = click.prompt("Node swap memory limit in GB, make sure it is larger then mem limit (press enter to use default)", type = int, default = 15)
```
Comment (dule1322) [inline coretex/cli/config.py:163]: Swap memory default should be set to RAM default times 2 (double the size of RAM)

File: coretex/cli/config.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    shm = click.prompt("Node shm limit in GB (press enter to use default)", type = int, default = 9)
```
Comment (dule1322) [inline coretex/cli/config.py:164]: Rename to `sharedMemory`. Message should print out `POSIX shared memory` instead of `shm`. Default should be 2 GB

File: coretex/cli/config.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    config["organizationID"] = organizationId
+    config["storagePath"] = storagePath
+    config["image"] = image
+
+    saveConfig(config)
+
+    generateDockerScript(image)
+    generateDockerCompose(nodeName, ram, swap, shm, organizationId)
+
+    click.echo("[Node Setup Done] Type coretex help for additional information")
```
Comment (dule1322) [inline coretex/cli/config.py:176]: coretex help should be enclosed in quotes

File: coretex/cli/config.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    click.echo("Important Information for Coretex Node service (command \"coretex node\"): [for using coretex node service you need to have Docker (https://www.docker.com/) installed and set up on your machine]")
+    click.echo("For automatic updates, if not already, make sure \"atd\" daemon is enabled. Either with \"ps -ef | grep atd\" or \"systemctl -a | grep atd\"")
+    click.echo("For additional help visit our documentation at https://docs.coretex.ai/v1/advanced/coretex-cli/troubleshooting")
+
+
+@click.command()
+@click.option("--user", is_flag = True, help = "Configure user settings")
+@click.option("--node", is_flag = True, help = "Configure node settings")
+def config(user: bool, node: bool) -> None:
+    if not (user ^ node):
```
Comment (dule1322) [inline coretex/cli/config.py:186]: Use `if not user and not node`

File: coretex/cli/project.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
+from typing import Optional
+
+import click
+from ..configuration import loadConfig, saveConfig
+from .. import Project, ProjectType
+
+
+def createProject(name: str) -> None:
+    try:
+        new_project = Project.createProject(name, ProjectType.other)
```
Comment (dule1322) [inline coretex/cli/project.py:10]: Just `project`. User should pick the type of the project which is to be created, it should not be hardcoded to `other`

File: coretex/cli/project.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        if new_project is not None:
+            config = loadConfig()
+            config["projectId"] = new_project.id
+            saveConfig(config)
```
Comment (dule1322) [inline coretex/cli/project.py:14]: This is a duplicate functionality used both here, and in `select` command. Extract it into a separate function

File: coretex/cli/project.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+@click.command()
+@click.option("--name", "-n", type = str, help = "Project name")
+@click.option("--id", type = int, help = "Project ID")
+def select(name: Optional[str], id: Optional[int]) -> None:
+    if name is None and id is None:
+        raise click.UsageError("Please use either --name / -n (for project name) or --id (for project ID)")
+
+    if name:
+        click.echo("Checking project name")
+        projects = Project.fetchAll(name = name, include_sessions = 1)
```
Comment (dule1322) [inline coretex/cli/project.py:28]: Use `Project.fetchOne` instead of `Project.fetchAll`. What is `include_sessions` parameter?

File: coretex/cli/project.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            click.echo(f"Could not find project with name \"{name}\"", err = True)
+            if click.confirm("Do you want to create a project by that name?", default = True):
+                createProject(name)
+        else:
+            click.echo(f"Project \"{name}\" selected successfully!")
+
+    if id:
+        click.echo("Checking project id")
+        project = Project.fetchById(id)
+        if project:
```
Comment (dule1322) [inline coretex/cli/project.py:39]: `Project.fetchById` returns an object, it never returns None, if object is not found an exception will be raised. Please handle appropriatelly.

File: coretex/configuration.py [hunk 12]
```diff
@@ -29,6 +29,7 @@ def getEnvVar(key: str, default: str) -> str:
 
     return os.environ[key]
 
+DEFAULT_CONFIG_PATH = Path.home() / ".config/coretex/config.json"
```
Comment (dule1322) [inline coretex/configuration.py:32]: Use `Path.home().joinpath("first", "second", "third", ...)`

File: coretex/cli/config.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+    """
+
+    dockerScript = CORETEX_CONFIG_DIR / "docker.sh"
+    with open(dockerScript, "w") as dockerFile:
+        dockerFile.write(dockerScriptContent)
+
+    os.chmod(dockerScript, 0o755)
+
+
+def generateDockerCompose(
```
Comment (dule1322) [inline coretex/cli/config.py:83]: Why is this needed?

File: coretex/cli/config.py [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+    config["storagePath"] = storagePath
+    config["token"] = response.getJson(dict).get("token")
+    config["refreshToken"] = response.getJson(dict).get("refresh_token")
+
+    saveConfig(config)
+
+    click.echo("User successfuly configured")
+
+
+def generateDockerScript(image: str) -> None:
```
Comment (dule1322) [inline coretex/cli/config.py:63]: Why is this needed?

File: coretex/cli/config.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+        if not click.prompt(
+            "User configuration already exists. Would you like to update (Y/n)?",
+            type = bool,
+            default = True,
+            show_default = False
+        ):
+            return
+
+    click.echo("Configuring user...")
+    while True:
```
Comment (dule1322) [inline coretex/cli/config.py:39]: Limit retires for authenticate to 3 times after which an error message will be printed out

File: coretex/cli/config.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+
+        click.echo("Failed to authenticate. Try again")
+
+    click.echo("Storage path should be the same as (if) used during --node config")
+    storagePath = click.prompt("Storage path (press enter to use default)", Path.home() / ".coretex", type = str)
+
+    config["username"] = username
+    config["password"] = password
+    config["storagePath"] = storagePath
+    config["token"] = response.getJson(dict).get("token")
```
Comment (dule1322) [inline coretex/cli/config.py:55]: `response.getJson` should only be called once.

File: coretex/cli/config.py [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+
+    config["nodeName"] = nodeName
+    config["organizationID"] = organizationId
+    config["storagePath"] = storagePath
+    config["image"] = image
+
+    saveConfig(config)
+
+    generateDockerScript(image)
+    generateDockerCompose(nodeName, ram, swap, shm, organizationId)
```
Comment (dule1322) [inline coretex/cli/config.py:174]: This is a part of starting the node, not configuring the node. Python has a docker library which will handle that for us, lets not use this approach.

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: coretex/cli/config.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+import click
+
+from ..configuration import loadConfig, saveConfig, isUserConfigured, isNodeConfigured
+from ..statistics import getAvilableRamMemory
+from ..networking import networkManager
+
+
+SERVER_URL = "dev.biomechservices.com"
+NODE_PORT = 5443
+CORETEX_CONFIG_DIR = Path.home() / ".config" / "coretex"
```
Comment (dule1322) [inline coretex/cli/config.py:16]: These variables are unused

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: coretex/statistics.py [hunk 19]
```diff
@@ -45,6 +45,17 @@ def getRamUsage() -> float:
     return psutil.virtual_memory().percent
 
 
+def getAvailableRamMemory() -> int:
+    """
+        Returns
+        -------
+        int -> total RAM memory as int
```
Comment (dule1322) [inline coretex/statistics.py:52]: In what scale? GB, MB, KB, etc...?

File: coretex/cli/config.py [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+    jsonResponse = response.getJson(dict)
+
+    return username, password, jsonResponse["token"], jsonResponse["refreshToken"]
+
+
+def registerNode(name: str) -> Optional[str]:
+    params = {
+        "machine_name": name
+    }
+    response = networkManager.post('service', params = params)
```
Comment (dule1322) [inline coretex/cli/config.py:36]: Use double quotes in Python. There is no need to pass `params` as a keyword parameter, it can be safely passed as a positional parameter.

File: coretex/cli/config.py [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+
+    if response.hasFailed():
+        click.echo("Failed to configure node. Please try again...")
+        return None
+
+    return str(response.getJson(dict)["access_token"])
```
Comment (dule1322) [inline coretex/cli/config.py:42]: `access_token` is already a string, why are you creating a new string object here?

File: coretex/cli/config.py [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+            "User configuration already exists. Would you like to update (Y/n)?",
+            type = bool,
+            default = True,
+            show_default = False
+        ):
+            return
+
+    click.echo("Configuring user...")
+    username = click.prompt("Email", type=str)
+    password = click.prompt("Password", type=str, hide_input=True)
```
Comment (dule1322) [inline coretex/cli/config.py:67]: Spacing is missing around `=` character

File: coretex/cli/project.py [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+                else:
+                    click.echo("Failed to select project. Please try again.")
+                    return
+
+    if id is not None:
+        click.echo ("Checking project id")
+        try:
+            project = Project.fetchById(id)
+            click.echo(f"Project with id \"{id}\" fetched successfully. Project name \"{project.name}\"")
+        except:
```
Comment (dule1322) [inline coretex/cli/project.py:68]: Handle appropriate exception here, not all raised exceptions

File: coretex/cli/utils.py [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+
+
+def arrowPrompt(choices: List[Any]) -> Any:
+    answers = inquirer.prompt([
+            inquirer.List("option",
+            message = "Use arrow keys to select an option",
+            choices = choices,
+            carousel = True,
+        )]
+    )
```
Comment (dule1322) [inline coretex/cli/utils.py:13]: Formatting is invalid

File: coretex/entities/project/project.py [hunk 25]
```diff
@@ -33,7 +35,7 @@ class Project(BaseObject):
     tasks: List[Task]
 
     @classmethod
-    def createProject(cls, name: str, projectType: ProjectType, description: Optional[str] = None) -> Optional[Self]:
+    def createProject(cls, name: str, projectType: ProjectType, description: Optional[str] = None) -> Self:
```
Comment (dule1322) [inline coretex/entities/project/project.py:38]: Update documentation as function signature changed and it is now raising an exception

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: coretex/cli/config.py (hunk 21 above)
Comment (bogdant36) [reply coretex/cli/config.py:42]: Regarding python response.getJson(dict)["access_token"] is type of Any

Comment (dule1322) [reply coretex/cli/config.py:42]: That might be what the linter is "guessing", but the actual `type(response.getJson(dict)["access_token"])` will be `str`. Perform validation instead of creating a new object here

File: coretex/cli/config.py [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+
+def registerNode(name: str) -> Optional[str]:
+    params = {
+        "machine_name": name
+    }
+    response = networkManager.post("service", params)
+
+    if response.hasFailed():
+        click.echo("Failed to configure node. Please try again...")
+        return None
```
Comment (dule1322) [inline coretex/cli/config.py:40]: raise an exception here instead of returning `None`

File: coretex/cli/config.py [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+
+from .utils import arrowPrompt
+from ..networking import networkManager
+from ..statistics import getAvailableRamMemory
+from ..configuration import loadConfig, saveConfig, isUserConfigured, isNodeConfigured
+
+
+def authenticate(retryCount: int = 0) -> Tuple[str, str, str, str]:
+    if retryCount >= 3:
+        raise click.ClickException("") #raise errror
```
Comment (dule1322) [inline coretex/cli/config.py:16]: raise `Exception` here with a proper message

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: coretex/cli/config.py [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+    }
+    response = networkManager.post("service", params)
+
+    if response.hasFailed():
+        click.echo("Failed to configure node. Please try again...")
+        return None
+
+    accessToken = response.getJson(dict).get("access_token")
+
+    if not isinstance(accessToken, str) and accessToken is not None:
```
Comment (dule1322) [inline coretex/cli/config.py:44]: There is no need for `None` check here, just `isinstance` is enough

**Review by dule1322 — APPROVED** [review]
